lib/ret_web/auth_channel.ex

```diff
@@ -0,0 +1,73 @@
+defmodule RetWeb.AuthChannel do
+  @moduledoc "Ret Web Channel for Authentication"
+
+  use RetWeb, :channel
+
+  alias Ret.{Statix, LoginToken, Account}
+
+  intercept(["auth_credentials"])
+
+  def join("auth:" <> _topic_key, _payload, socket) do
+    # Expire channel in 5 minutes
+    Process.send_after(self(), :channel_expired, 60 * 1000 * 5)
+
+    # Rate limit joins to reduce attack surface
+    :timer.sleep(500)
+
+    Statix.increment("ret.channels.auth.joins.ok")
+    {:ok, "{}", socket}
+  end
+
+  def handle_in("auth_request", %{"email" => email}, socket) do
+    if !Map.get(socket.assigns, :used) do
+      socket = socket |> assign(:used, true)
+
+      # Create token + send email
+      token = LoginToken.new_token_for_email(email)
+      signin_args = %{topic: socket.topic, token: token}
+
+      RetWeb.Email.auth_email(email, signin_args) |> Ret.Mailer.deliver_now()
```

Comment:
Is mail rate-limited somewhere?

Reply:
the channel joins are rate limited and you can only send one of these per topic. but its still a pretty high limit (2 TPS) and obviously you can open many sockets simultaneously. as far as globally rate limiting emails I'm not sure if there's a knob on our side, I can check. SES almost certainly will cut us off (but obviously that would suck)

Reply:
we're currently using `plug_attack` to rate limit certain requests by IP, but that doesn't seem to expose a generic ETS based rate limiting API. It seems like we should explore this alternative which would let us rate limit HTTP requests or anything else:

https://github.com/ExHammer/hammer

I'll create an issue for it